tools/congestion-model/src/workload/fairness_benchmark.rs

```diff
@@ -0,0 +1,84 @@
+use crate::{GGas, ReceiptDefinition, ShardId, TransactionBuilder, GAS_LIMIT, TGAS};
+
+use super::Producer;
+
+/// Transaction producer that tests fairness for shards.
+///
+/// One shard has to execute receipts from all others. But it is also trying to
+/// send new transactions to all others.
+///
+/// A fair strategy should allow to send the busy shard at least some work to
+/// other shards, since those are not above capacity. Thus, by looking at
+/// utilization, we can get a measure of fairness between shards.
+pub struct FairnessBenchmarkProducer {
+    pub receipt_size: u64,
+    pub attached_gas: GGas,
+    pub execution_gas: GGas,
+    pub conversion_gas: GGas,
+}
+
+impl Producer for FairnessBenchmarkProducer {
+    fn init(&mut self, _shards: &[ShardId]) {}
+
+    fn produce_transactions(
+        &mut self,
+        _round: crate::Round,
+        shards: &[ShardId],
+        tx_factory: &mut dyn FnMut(ShardId) -> TransactionBuilder,
+    ) -> Vec<TransactionBuilder> {
+        let mut out = vec![];
+
+        let busy_shard = shards[0];
+        let other_shards = &shards[1..];
+
+        let mut gas_to_busy = 0;
+        let mut gas_to_non_busy = 0;
+
+        // Send from all to the busy shard until it receives at least 4 times
+        // what it can handle
+        while gas_to_busy < GAS_LIMIT * 4 {
+            for &other_shard in other_shards {
+                let mut tx_to_busy_shard = tx_factory(other_shard);
+                self.produce_tx(busy_shard, &mut tx_to_busy_shard);
+                out.push(tx_to_busy_shard);
+                gas_to_busy += self.execution_gas;
+            }
+        }
+        
+        // Send from the busy shard to all other shards, using around 90% of
+        // their capacity.
+        while gas_to_non_busy < GAS_LIMIT * 10 / 9 {
```

Comment:
based on this I would expect the maximum achievable utilization to be around 1000Tgas * (1 + 0.9 * (n - 1))

Reply:
this only accounts for execution gas, not transaction -> receipt conversion gas

Reply:
but the conversion is only 1Tgas, that's like extra 1%

Reply:
Yeah, and that's almost enough to explain it. Combined with a small bug I just noticed thanks to you :) It wanted it to be 90% but it was 10/9 = 111%. Let me fix that real quick... Maybe just < GAS_LIMIT instead of 90% seems fine, it should still not cause congestion in an ideal strategy.

1015 TGas on average per shard means 15 converted receipts per chunk, on average.
I ran it with 8 shards, so it would be 120 receipts per round.

The busy shard should send 9 receipts to 7 other shards = 63 transactions.
And the other 7 shards send combined at least 40 receipts to the busy shard. I think they send 7x6 = 42.
Then it's 105 transactions per block, or ~13 per chunk.